File: sdk/go/shared.go
package cutctx

import (
	"fmt"
	"sort"
	"sync"
)

// SharedContext provides thread-safe key-value storage for sharing
// context across multiple operations.
type SharedContext struct {
	mu    sync.RWMutex
	items map[string]string
}

// SharedStats holds statistics about the shared context.
type SharedStats struct {
	Entries int      `json:"entries"`
	Keys    []string `json:"keys"`
}

// NewSharedContext creates a new SharedContext.
func NewSharedContext() *SharedContext {
	return &SharedContext{
		items: make(map[string]string),
	}
}

// Put stores a key-value pair in the shared context.
func (sc *SharedContext) Put(key, value string) {
	sc.mu.Lock()
	defer sc.mu.Unlock()
	sc.items[key] = value
}

// Get retrieves a value by key. Returns the value and whether it was found.
func (sc *SharedContext) Get(key string) (string, bool) {
	sc.mu.RLock()
	defer sc.mu.RUnlock()
	val, ok := sc.items[key]
	return val, ok
}

// List returns a copy of all entries in the shared context.
func (sc *SharedContext) List() map[string]string {
	sc.mu.RLock()
	defer sc.mu.RUnlock()
	result := make(map[string]string, len(sc.items))
	for k, v := range sc.items {
		result[k] = v
	}
	return result
}

// Clear removes all entries from the shared context.
func (sc *SharedContext) Clear() {
	sc.mu.Lock()
	defer sc.mu.Unlock()
	sc.items = make(map[string]string)
}

// Stats returns statistics about the shared context.
func (sc *SharedContext) Stats() SharedStats {
	sc.mu.RLock()
	defer sc.mu.RUnlock()
	keys := make([]string, 0, len(sc.items))
	for k := range sc.items {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return SharedStats{
		Entries: len(sc.items),
		Keys:    keys,
	}
}

// String returns a human-readable representation.
func (sc *SharedContext) String() string {
	stats := sc.Stats()
	return fmt.Sprintf("SharedContext{entries=%d, keys=%v}", stats.Entries, stats.Keys)
}

File: sdk/python/cutctx_sdk/shared.py
"""Thread-safe shared context for key-value storage."""

from __future__ import annotations

import threading
from typing import Any


class SharedContext:
    """Thread-safe key-value store for sharing context across operations.

    Example:
        ctx = SharedContext()
        ctx.put("project", "my-app")
        ctx.get("project")  # ("my-app", True)
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._items: dict[str, str] = {}

    def put(self, key: str, value: str) -> None:
        """Store a key-value pair."""
        with self._lock:
            self._items[key] = value

    def get(self, key: str) -> tuple[str, bool]:
        """Retrieve a value by key. Returns (value, found)."""
        with self._lock:
            val = self._items.get(key)
            return (val, val is not None)

    def list(self) -> dict[str, str]:
        """Return a copy of all entries."""
        with self._lock:
            return dict(self._items)

    def clear(self) -> None:
        """Remove all entries."""
        with self._lock:
            self._items.clear()

    def stats(self) -> dict[str, Any]:
        """Return statistics about the shared context."""
        with self._lock:
            keys = sorted(self._items.keys())
            return {
                "entries": len(self._items),
                "keys": keys,
            }

    def __repr__(self) -> str:
        stats = self.stats()
        return f"SharedContext(entries={stats['entries']}, keys={stats['keys']!r})"
